File: branches/convenience-views/implementation/UrakawaToolkit/urakawa/properties/xml/XmlPropertyElementNameVisitor.cs
using System;
using urakawa.core;
using urakawa.core.visitor;

namespace urakawa.properties.xml
{
	/// <summary>
	/// Summary description for XmlPropertyElementNameVisitor.
	/// </summary>
	public class XmlPropertyElementNameVisitor : ICoreNodeVisitor
	{
		private System.Collections.Generic.IList<string> mNamesToMatch;
		private System.Collections.Generic.IList<ICoreNode> mNodes;
		private Type mXmlPropertyType;

		/// <summary>
		/// The constructor
		/// </summary>
		public XmlPropertyElementNameVisitor()
		{
			mNamesToMatch = new System.Collections.Generic.List<string>();
			mNodes = new System.Collections.Generic.List<ICoreNode>();
			mXmlPropertyType = typeof(XmlProperty);
		}

		public void setXmlPropertyType(Type newType)
		{
			if (typeof(IXmlProperty).IsAssignableFrom(newType))
			{
				throw new exception.MethodParameterIsWrongTypeException(
					"The new xml property type must a Type that can be assigned to an IXmlProperty");
			}
		}

		/// <summary>
		/// Add an element name to the collection of search terms.  
		/// The search terms should be considered an "OR"-list.
		/// </summary>
		/// <param name="localName">The local name part of the element name</param>
		/// <param name="namespaceUri">The namespace uri part of the element name</param>
		public void addElementName(string localName, string namespaceUri)
		{
			mNamesToMatch.Add(String.Format("{0}:{1}", namespaceUri, localName));
		}



		/// <summary>
		/// Get the results of the tree visit to see if any nodes were found
		/// whose XML properties matched the search request.
		/// </summary>
		/// <returns>The list</returns>
		public System.Collections.Generic.IList<ICoreNode> getResults()
		{
			return mNodes;
		}

		private bool isMatch(string localName, string namespaceUri)
		{
			return mNamesToMatch.Contains(String.Format("{0}:{1}", localName, namespaceUri));
		}
		#region ICoreNodeVisitor Members

		/// <summary>
		/// look at the current node and see if it has an XML property 
		/// that is interesting to us.  if so, add it to our internal list.
		/// </summary>
		/// <param name="node"></param>
		/// <returns></returns>
		public bool preVisit(ICoreNode node)
		{
			IXmlProperty xp = (IXmlProperty)node.getProperty(mXmlPropertyType);

			if (xp != null && isMatch(xp.getLocalName(), xp.getNamespaceUri()) == true)
			{
				mNodes.Add(node);
			}
			
			return true;
		}

		/// <summary>
		/// this visitor does nothing post-visit
		/// </summary>
		/// <param name="node"></param>
		public void postVisit(ICoreNode node)
		{
			//empty
		}

		#endregion
	}
}

File: trunk/java/core/src/org/daisy/urakawa/core/visitor/examples/XmlPropertyElementNameVisitor.java
package org.daisy.urakawa.core.visitor.examples;

import java.util.LinkedList;
import java.util.List;

import org.daisy.urakawa.core.ITreeNode;
import org.daisy.urakawa.core.visitor.ITreeNodeVisitor;
import org.daisy.urakawa.exception.IsNotInitializedException;
import org.daisy.urakawa.exception.MethodParameterIsNullException;
import org.daisy.urakawa.exception.MethodParameterIsWrongTypeException;
import org.daisy.urakawa.property.xml.IXmlProperty;

/**
 *
 */
public class XmlPropertyElementNameVisitor implements ITreeNodeVisitor
{
    private List<String> mNamesToMatch;
    private List<ITreeNode> mNodes;
    @SuppressWarnings("unused")
    private Class<IXmlProperty> mXmlPropertyType = IXmlProperty.class;

    /**
	 * 
	 */
    public XmlPropertyElementNameVisitor()
    {
        mNamesToMatch = new LinkedList<String>();
        mNodes = new LinkedList<ITreeNode>();
        mXmlPropertyType = IXmlProperty.class;
    }

    /**
     * @param newType
     * @throws MethodParameterIsWrongTypeException
     */
    public void setXmlPropertyType(Class<IXmlProperty> newType)
            throws MethodParameterIsWrongTypeException
    {
        if (!IXmlProperty.class.isAssignableFrom(newType))
        {
            throw new MethodParameterIsWrongTypeException();
        }
        mXmlPropertyType = newType;
    }

    /**
     * @param localName
     * @param namespaceUri
     */
    public void addElementName(String localName, String namespaceUri)
    {
        mNamesToMatch.add(String.format("{0}:{1}", namespaceUri, localName));
    }

    /**
     * @return list
     */
    public List<ITreeNode> getResults()
    {
        return mNodes;
    }

    private boolean isMatch(String localName, String namespaceUri)
    {
        return mNamesToMatch.contains(String.format("{0}:{1}", namespaceUri,
                localName));
    }

    public void postVisit(ITreeNode node) throws MethodParameterIsNullException
    {
        /**
         * Does nothing.
         */
    }

    public boolean preVisit(ITreeNode node)
            throws MethodParameterIsNullException
    {
        IXmlProperty xp = node.<IXmlProperty> getProperty(IXmlProperty.class);
        try
        {
            if (xp != null
                    && isMatch(xp.getLocalName(), xp.getNamespace()) == true)
            {
                mNodes.add(node);
            }
        }
        catch (IsNotInitializedException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
        return true;
    }
}
